### DISCUSSION TITLE
obtain atom mapping used in CalcRMS

### DISCUSSION THREAD

**Discussion by fabioatuib** [rdkit/rdkit#4088] [Q&A] [answered]
Hello!
Is it possible to obtain the atom mapping that corresponds to the rmsd value CalcRMS outputs? If so, how?

Thank you!

**Comment by greglandrum** [comment, marked as answer]
There's currently no way to do that with the `CalcRMS()` function itself, but you can create a function for it.
Here's a very basic example of how that might look:
```
def CalcRMSAndMatch(prbMol,refMol,prbId=-1,refId=-1):
    matches = refMol.GetSubstructMatches(prbMol,uniquify=False)
    bestRMS=1e9
    for match in matches:
        mmap = list(zip(match,range(len(match))))
        rms = AllChem.CalcRMS(prbMol,refMol,prbId,refId,[mmap])
        if rms<bestRMS:
            bestRMS=rms
            bestMatch = tuple(match)
    return bestRMS,bestMatch
```
and what the results look like:
```
In [25]: m = Chem.AddHs(Chem.MolFromSmiles('c1ccccc1Cc1ccccc1'))

In [26]: AllChem.EmbedMultipleConfs(m,10,randomSeed=0xf00d)
Out[26]: <rdkit.rdBase._vecti at 0x7f5f40f640b0>

In [27]: CalcRMSAndMatch(m,m,0,1)
Out[27]: 
(2.839441526383006,
 (12,
  11,
  10,
  9,
  8,
  7,
  6,
  5,
  0,
  1,
  2,
  3,
  4,
  24,
  23,
  22,
  21,
  20,
  19,
  18,
  13,
  14,
  15,
  16,
  17))
```

**Reply by fabioatuib** [reply]
Thank you for the suggestion! 
Follow up question:
What is it that uniquify=False does (I don't understand the definition in the docs: "determines whether or not the matches are uniquified")?

**Reply by ptosco** [reply]
A couple of simple example should make it clear:
```
from rdkit import Chem
query = Chem.MolFromSmiles("Cc1cc(C)ccc1")

query
```
![image](https://user-images.githubusercontent.com/5244385/116532599-b489c300-a8e0-11eb-9010-0a2782f03080.png)
```
query.GetSubstructMatches(query)
((0, 1, 2, 3, 4, 5, 6, 7),)

query.GetSubstructMatches(query, uniquify=False)
((0, 1, 2, 3, 4, 5, 6, 7), (4, 3, 2, 1, 0, 7, 6, 5))

mol = Chem.MolFromSmiles("Cc1cc(C)cc(C)c1")
mol
```
![image](https://user-images.githubusercontent.com/5244385/116533030-3548bf00-a8e1-11eb-948c-c1d5aabb0b68.png)
```
mol.GetSubstructMatches(query)
((0, 1, 2, 3, 4, 5, 6, 8), (0, 1, 8, 6, 7, 5, 3, 2), (4, 3, 5, 6, 7, 8, 1, 2))

mol.GetSubstructMatches(query, uniquify=False)
((0, 1, 2, 3, 4, 5, 6, 8),
 (0, 1, 8, 6, 7, 5, 3, 2),
 (4, 3, 2, 1, 0, 8, 6, 5),
 (4, 3, 5, 6, 7, 8, 1, 2),
 (7, 6, 5, 3, 4, 2, 1, 8),
 (7, 6, 8, 1, 0, 2, 3, 5))
```

With `uniquify=True` you get a single match involving a certain tuple of indices.
With `uniquify=False` you get all possible symmetry-equivalent matches involving a certain tuple of indices.

**Reply by greglandrum** [reply]
the default behavior is to only return results which include unique atom IDs (that's uniquify=True). If you want to find all the mappings between a molecule and itself you have to turn this off.

Please note that the code I shared, like `CalcRMS()` itself, can take a very, very long time if moleules have hydrogen atoms on them (every  CH3 group triples the number of mappings, every CH2 doubles it). You probably should remove Hs before calling any of those functions.
